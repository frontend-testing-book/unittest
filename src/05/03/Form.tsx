type Props = {
  name: string
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void
}
export const Form = ({ name, onSubmit }: Props) => {
  return (
    <form
      onSubmit={(event) => {
        event.preventDefault()
        onSubmit?.(event)
      }}
    >
      <h2>アカウント情報</h2>
      <p>{name}</p>
      <div>
        <button>編集する</button>
      </div>
    </form>
  )
}
