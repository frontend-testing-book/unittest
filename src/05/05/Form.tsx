import { useId, useState } from 'react'

import { Agreement } from './Agreement'
import { InputAccount } from './InputAccount'

export const Form = () => {
  const [checked, setChecked] = useState(false)
  const headingId = useId()
  return (
    <form aria-labelledby={headingId}>
      <h2 id={headingId}>新規アカウント登録</h2>
      <InputAccount />
      <Agreement
        onChange={(event) => {
          setChecked(event.currentTarget.checked)
        }}
      />
      <div>
        <button disabled={!checked}>サインアップ</button>
      </div>
    </form>
  )
}
