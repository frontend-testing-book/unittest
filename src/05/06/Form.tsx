import { useState } from 'react'

import { ContactNumber } from './ContactNumber'
import { DeliveryAddress } from './DeliveryAddress'
import { PastDeliveryAddress } from './PastDeliveryAddress'
import { RegisterDeliveryAddress } from './RegisterDeliveryAddress'

export type AddressOption = React.ComponentProps<'option'> & { id: string }
export type Props = {
  deliveryAddresses?: AddressOption[]
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void
}
export const Form = (props: Props) => {
  const [registerNew, setRegisterNew] = useState<boolean | undefined>(undefined)
  return (
    <form onSubmit={props.onSubmit}>
      <h2>お届け先情報の入力</h2>
      <ContactNumber />
      {props.deliveryAddresses?.length ? (
        <>
          <RegisterDeliveryAddress onChange={setRegisterNew} />
          {registerNew ? (
            <DeliveryAddress title="新しいお届け先" />
          ) : (
            <PastDeliveryAddress
              disabled={registerNew === undefined}
              options={props.deliveryAddresses}
            />
          )}
        </>
      ) : (
        <DeliveryAddress />
      )}
      <hr />
      <div>
        <button>注文内容の確認へ進む</button>
      </div>
    </form>
  )
}
