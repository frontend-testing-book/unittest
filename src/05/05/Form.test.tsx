import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { Form } from './Form'

const user = userEvent.setup()

test('form のアクセシブルネームは、見出しを引用している', () => {
  render(<Form />)
  expect(
    screen.getByRole('form', { name: '新規アカウント登録' }),
  ).toBeInTheDocument()
})

test('主要エリアが表示されている', () => {
  render(<Form />)
  expect(
    screen.getByRole('heading', { name: '新規アカウント登録' }),
  ).toBeInTheDocument()
  expect(
    screen.getByRole('group', { name: 'アカウント情報の入力' }),
  ).toBeInTheDocument()
  expect(
    screen.getByRole('group', { name: '利用規約の同意' }),
  ).toBeInTheDocument()
  expect(
    screen.getByRole('button', { name: 'サインアップ' }),
  ).toBeInTheDocument()
})

test('「サインアップ」ボタンは非活性', () => {
  render(<Form />)
  expect(screen.getByRole('button', { name: 'サインアップ' })).toBeDisabled()
})

test('「利用規約の同意」チェックボックスを押下すると「サインアップ」ボタンは活性化', async () => {
  render(<Form />)
  await user.click(screen.getByRole('checkbox'))
  expect(screen.getByRole('button', { name: 'サインアップ' })).toBeEnabled()
})

test('Snapshot: 新規アカウント登録フォームが表示される', () => {
  const { container } = render(<Form />)
  expect(container).toMatchSnapshot()
})
