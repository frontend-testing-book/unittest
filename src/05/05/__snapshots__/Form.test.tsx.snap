// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: 新規アカウント登録フォームが表示される 1`] = `
<div>
  <form
    aria-labelledby=":r4:"
  >
    <h2
      id=":r4:"
    >
      新規アカウント登録
    </h2>
    <fieldset>
      <legend>
        アカウント情報の入力
      </legend>
      <div>
        <label>
          メールアドレス
          <input
            placeholder="example@test.com"
            type="text"
          />
        </label>
      </div>
      <div>
        <label>
          パスワード
          <input
            placeholder="8文字以上で入力"
            type="password"
          />
        </label>
      </div>
    </fieldset>
    <fieldset>
      <legend>
        利用規約の同意
      </legend>
      <label>
        <span>
          <input
            name="agree"
            style="display: none;"
            type="checkbox"
          />
          <span
            style="display: inline-block; text-align: center; width: 24px; height: 24px; border-radius: 4px; cursor: pointer; color: rgb(255, 255, 255); background-color: rgb(204, 204, 204);"
          >
            ✔︎
          </span>
        </span>
        当サービスの
        <a
          href="/terms"
        >
          利用規約
        </a>
        を確認し、これに同意します
      </label>
    </fieldset>
    <div>
      <button
        disabled=""
      >
        サインアップ
      </button>
    </div>
  </form>
</div>
`;
